### ISSUE TITLE
Non deterministic output from pretrained ELMo

### ISSUE THREAD

**Issue by atremblay** [allenai/allennlp#2516]
**Describe the bug**
Non-deterministic output from the ELMo embedding

**To Reproduce**
Following this tutorial
https://github.com/allenai/allennlp/blob/master/tutorials/how_to/elmo.md#using-elmo-as-a-pytorch-module-to-train-a-new-model

``` python
from allennlp.modules.elmo import Elmo, batch_to_ids

options_file = "https://s3-us-west-2.amazonaws.com/allennlp/models/elmo/2x4096_512_2048cnn_2xhighway/elmo_2x4096_512_2048cnn_2xhighway_options.json"
weight_file = "https://s3-us-west-2.amazonaws.com/allennlp/models/elmo/2x4096_512_2048cnn_2xhighway/elmo_2x4096_512_2048cnn_2xhighway_weights.hdf5"

# Compute two different representation for each token.
# Each representation is a linear weighted combination for the
# 3 layers in ELMo (i.e., charcnn, the outputs of the two BiLSTM))
elmo = Elmo(options_file, weight_file, 1, dropout=0)
elmo.eval()

# use batch_to_ids to convert sentences to character ids
sentences = [['First', 'sentence', '.']]
character_ids = batch_to_ids(sentences)

first = elmo(character_ids)['elmo_representations'][0]
second = elmo(character_ids)['elmo_representations'][0]

>>> (first - second)  # =/= 0 everywhere
```
**Expected behavior**
Having a dropout of 0 and being in eval mode, I would have expected `first` and `second` to be identical.

**System (please complete the following information):**
 - OS: Ubuntu 16.04
 - Python version: 3.6.7
 - AllenNLP version: 0.7.2
 - PyTorch version: 0.4.1

**Comment by HarshTrivedi** [comment]
You can find the reason [here](https://github.com/allenai/allennlp/blob/master/tutorials/how_to/elmo.md#notes-on-statefulness-and-non-determinism).

**Comment by atremblay** [comment]
Thanks for the super fast response. I saw that running the same sentence a couple of times stabilized the output, but having to do that for every piece of texts is not really efficient. 
I now have a doubt now that I may have misunderstood how `batch_to_ids` works. I thought it was going to standardize everything to fit ELMo implementation requirements. Should I be adding <S> and </S> manually as tokens?

**Comment by schmmd** [comment]
@atremblay once the model is "warmed up" you should get "more consistent" results.  We recently trained a model that uses a transformer that doesn't have this unfortunate issue, but I'm not sure whether we've uploaded that yet to allennlp.org/elmo--@brendanr?

**Comment by pkubik** [comment]
Couldn't it be solved by simply capturing the final states of a warmed-up instance of the model and assign it to the corresponding initial states of each RNNs? We could just store those vectors on persistent storage and load them along with the model's weights. We would avoid both the warm-up issue and the results inconsistency.

Of course one could try to find the optimal initial RNN states, but it seems like it doesn't significantly affect the scores.

**Comment by andreaferretti** [comment]
One thing is not clear to me reading [this](https://github.com/allenai/allennlp/blob/master/tutorials/how_to/elmo.md#notes-on-statefulness-and-non-determinism):

> It is important to always add the <S> and </S> tokens to each sentence. The allennlp code handles this behind the scenes, but if you are handing padding and indexing in a different manner then take care to ensure this is handled appropriately.

I am not sure **which** allennlp code handles this. For instance, is this concern handled by the `ElmoTokenEmbedder`?

I am doing something more or less like this (simplified from real code)

```python
options_file = os.path.join(options.elmo_dir, f'elmo_{elmo_model}_options.json')
weight_file = os.path.join(options.elmo_dir, f'elmo_{elmo_model}_weights.hdf5')
elmo_embedder = ElmoTokenEmbedder(options_file, weight_file)
word_embeddings = BasicTextFieldEmbedder({'tokens': elmo_embedder})
elmo_token_indexer = ELMoTokenCharactersIndexer()
...
tokens = [Token(word) for word in sentence]
instance = Instance({
    'tokens': TextField(tokens, token_indexers={'tokens': elmo_token_indexer})
})
...
# later in the model I receive the tokens from the above instance
embeddings = word_embeddings(tokens)
# more processing
```

I am seeing wildly variable output from the model, from run to run. If I change the tokens to

```python
tokens = [Token('<S>')] + [Token(word) for word in sentence] + [Token('</S>')]
```

I still see wildly variable output.

I understand that I should warm up the model, but I am not sure whether the version with `<S>` and `</S>` added is correct, or if `ElmoTokenEmbedder` takes care of this.

Moreover, should I do this during training as well? In that case, I am getting instances directly from the `StanfordSentimentTreeBankDatasetReader`. Does *this* class handle the adjunction of `<S>` and `</S>` tokens?

**Comment by brendan-ai2** [comment]
Transformer ELMo is available here: https://s3-us-west-2.amazonaws.com/allennlp/models/transformer-elmo-2019.01.10.tar.gz You can find a guide to using (and retraining) it at https://github.com/allenai/allennlp/blob/master/tutorials/how_to/training_transformer_elmo.md#using-transformer-elmo-with-existing-allennlp-models.

**Comment by brendan-ai2** [comment]
@andreaferretti, yes the `ElmoTokenEmbedder` handles this. (The actual addition is somewhat buried. See https://github.com/allenai/allennlp/blob/master/allennlp/modules/elmo.py#L267 if you're curious.) It also by default removes the embeddings corresponding to the added `<S>` and `</S>` so that your model can work with sentences of the length that you'd expect.

I'm not familiar with the `StanfordSentimentTreeBankDatasetReader` myself, but I'd suggest simply calling its `read` method and inspecting some of the returned instances to see if they contain the special tokens. But to be clear, if you're using it with the `ElmoTokenEmbedder` (with the default settings) it shouldn't need or take the special tokens.

With respect to the varying output I'd suggest either using the Transformer ELMo I linked above or structuring your experiments such that you're resilient to the variability in standard ELMo, i.e. perform multiple runs with different random seeds (set these in your config https://github.com/allenai/allennlp/blob/2b52492e8002a534e647495ff55d592589a7d930/allennlp/common/util.py#L193) and consider the variance. Well, that's ideal for Transformer ELMo too, of course. :)

**Comment by andreaferretti** [comment]
@brendan-ai2 thank you!

**Comment by weiguowilliam** [comment]
> utput I'd suggest either using the Transformer ELMo I linked above or structuring your experiments such that you're resilient to the variability in standard ELMo, i.e. perform multiple runs with different ra

I used the TransformerElmo link you provided above, but it is still inconsistent.

**Comment by brendanr** [comment]
@weiguowilliam, that's unexpected. I'd suggest you open a new issue with
detailed information about how to reproduce the error (and what precisely
it is.)

On Mon, Feb 24, 2020 at 2:57 PM weiguowilliam <notifications@github.com>
wrote:
>
> utput I'd suggest either using the Transformer ELMo I linked above or
structuring your experiments such that you're resilient to the variability
in standard ELMo, i.e. perform multiple runs with different ra
>
> I used the TransformerElmo link you provided above, but it is still
inconsistent.

**Comment by weiguowilliam** [comment]
> @weiguowilliam, that's unexpected. I'd suggest you open a new issue with detailed information about how to reproduce the error (and what precisely it is.)
> On Mon, Feb 24, 2020 at 2:57 PM weiguowilliam ***@***.***> wrote: utput I'd suggest either using the Transformer ELMo I linked above or
> structuring your experiments such that you're resilient to the variability in standard ELMo, i.e. perform multiple runs with different ra
> I used the TransformerElmo link you provided above, but it is still
> inconsistent.
> […](#)
> — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or unsubscribe.

I opened a new issue here:
https://github.com/allenai/allennlp/issues/2516#issuecomment-590597305

**Comment by weiguowilliam** [comment]
> @weiguowilliam, that's unexpected. I'd suggest you open a new issue with detailed information about how to reproduce the error (and what precisely it is.)
> On Mon, Feb 24, 2020 at 2:57 PM weiguowilliam ***@***.***> wrote: utput I'd suggest either using the Transformer ELMo I linked above or
> structuring your experiments such that you're resilient to the variability in standard ELMo, i.e. perform multiple runs with different ra
> I used the TransformerElmo link you provided above, but it is still
> inconsistent.
> […](#)
> — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or unsubscribe.

Thank you. I have opened a new issue here:
https://github.com/allenai/allennlp/issues/3845
